include/materials/norsand.tcc

```diff
@@ -375,19 +378,44 @@ void mpm::NorSand<Tdim>::compute_plastic_tensor(const Vector6d& stress,
   dj3_dsigma(0) = dev1.dot(dev1) - (2. / 3.) * j2;
   dj3_dsigma(1) = dev2.dot(dev2) - (2. / 3.) * j2;
   dj3_dsigma(2) = dev3.dot(dev3) - (2. / 3.) * j2;
-  dj3_dsigma(3) = dev1.dot(dev2);
-  dj3_dsigma(4) = dev2.dot(dev3);
-  dj3_dsigma(5) = dev1.dot(dev3);
+  dj3_dsigma(3) = 2.0 * dev1.dot(dev2);
+  dj3_dsigma(4) = 2.0 * dev2.dot(dev3);
+  dj3_dsigma(5) = 2.0 * dev1.dot(dev3);
 
   if (Tdim == 2) {
     dj3_dsigma(4) = 0.;
     dj3_dsigma(5) = 0.;
   }
 
+  // Declare R as zero to avoid division by zero J2
+  double R = 0.0;
+
+  // Define derivatives of R in terms of J2 and J3
+  double dR_dj2 = -9.0 / 4.0 * sqrt(3.0) * j3;
+  double dR_dj3 = 3.0 / 2.0 * sqrt(3.0);
+
+  // Compute derivative of theta in terms of R
+  double dtheta_dR = 1.0 / 3.0;
+
+  // Update when J2 is non zero
+  if (abs(j2) > 1.0E-6) {
+    // Update R
+    R = j3 / 2.0 * pow(j2 / 3.0, -1.5);
+    // Update derivatives of R
+    dR_dj2 *= pow(j2, -2.5);
+    dR_dj3 *= pow(j2, -1.5);
+    // Update derivative of theta in terms of R, check for sqrt of zero
+    if (abs(1 - pow(R, 2.0)) < 1.0E-6) {
```

Comment:
Integer exponent is way faster than double or float. Please use integer if possible.
```suggestion
    if (abs(1 - pow(R, 2)) < 1.0E-6) {
```